packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -1592,41 +1624,60 @@ namespace Ifpack2 {
       }
 
     public:
-      
+
+      struct ExtractAndFactorizeScalarTag {};
+      struct ExtractAndFactorizeInternalVectorTag {};
+
       KOKKOS_INLINE_FUNCTION 
       void 
-      operator() (const member_type &member) const {
+      operator() (const ExtractAndFactorizeScalarTag &, const member_type &member) const {
 	// btdm is packed and sorted from largest one 
 	const local_ordinal_type packidx = member.league_rank();
 
 	const local_ordinal_type partidx = packptr(packidx);
 	const local_ordinal_type npacks = packptr(packidx+1) - partidx;
 	const local_ordinal_type i0 = pack_td_ptr(partidx);
 	const local_ordinal_type nrows = partptr(partidx+1) - partptr(partidx);
+	Kokkos::parallel_for
+	  (Kokkos::ThreadVectorRange(member, vector_length_value), [&](const local_ordinal_type &v) {
+	    if (v < npacks) extract(member, partidx+v, v);
+	    factorize(member, i0, nrows, v, scalar_values);
+	  });
+      }
 
+      KOKKOS_INLINE_FUNCTION 
+      void 
+      operator() (const ExtractAndFactorizeInternalVectorTag &, const member_type &member) const {
+	// btdm is packed and sorted from largest one 
+	const local_ordinal_type packidx = member.league_rank();
+
+	const local_ordinal_type partidx = packptr(packidx);
+	const local_ordinal_type npacks = packptr(packidx+1) - partidx;
+	const local_ordinal_type i0 = pack_td_ptr(partidx);
+	const local_ordinal_type nrows = partptr(partidx+1) - partptr(partidx);
 	Kokkos::parallel_for
 	  (Kokkos::ThreadVectorRange(member, vector_loop_size), [&](const local_ordinal_type &v) {
             const local_ordinal_type vbeg = v*internal_vector_length;
             const local_ordinal_type vend = vbeg + internal_vector_length;
             for (local_ordinal_type vv=vbeg;vv<vend;++vv)
-              if (vv < npacks) extract(member, partidx+vv, vv);
-	    member.team_barrier();
-	    factorize(member, i0, nrows, v);
+	    if (vv < npacks) extract(member, partidx+vv, vv);
+	    factorize(member, i0, nrows, v, internal_vector_values);
```

Comment:
See above. Does one need a team barrier here ? Just making sure.

Reply:
In practice, considering the thread mapping to matrices, it is safe to remove the barrier. In the LU factorization, the first element is mapped to the same thread and after first row is processed, it has a team barrier.